### PR TITLE
rfc: Preload Realm for Service Workers

### PR DESCRIPTION
**Pull request by samuelmaddock**
Reopened from #4 as my fork somehow became unlinked

> Preload realms provide an isolated context to run JavaScript prior to code evaluating in another
context. This expands functionality of Electron's preload scripts to provide new context targets.
This proposal intends to target Service Worker contexts with potential plans for new contexts (web
workers, shared workers) in future proposals.

[Read rendered document](https://github.com/samuelmaddock/rfcs/blob/preload-realms-rfc/text/0002-preload-realm.md)

Recommended prerequisite readings:
- [Design of V8 bindings](https://chromium.googlesource.com/chromium/src/third_party/+/master/blink/renderer/bindings/core/v8/V8BindingDesign.md)
- [ShadowRealms explainer](https://github.com/tc39/proposal-shadowrealm/blob/main/explainer.md)

If you're curious, there's a very rough wip prototype being written in https://github.com/samuelmaddock/electron/tree/feat/preload-realm

-----

Todo:
- [x] Try to get a better understanding of where ShadowRealm is at in the TC39 process.
- [x] Call out the option of landing the API as experimental until ShadowRealm is stable

### CODE REVIEW

**Comment by samuelmaddock** [comment]
Recent changes:
- `ipcMain` no longer receives SW IPC messages in favor of handling them on `ServiceWorkerMain.ipc`.
- "Main world" is now used by contextBridge instead of the added "initiator world" to simplify reusing preload scripts between frames and service workers.

**Review by itsananderson — APPROVED** [review]
Since the ShadowRealms proposal is apparently still in stage 2.7, we should make sure we mark this API as experimental until the web platform API is fully locked in, but otherwise this RFC LGTM

File: text/0002-preload-realm.md [hunk 1]
```diff
@@ -0,0 +1,482 @@
+- Start Date: 2024-01-11
+- RFC PR: [electron/rfcs#4](https://github.com/electron/rfcs/pull/4)
```
Comment (erickzhao) [inline text/0002-preload-realm.md:2]:
```suggestion
- RFC PR: [electron/rfcs#4](https://github.com/electron/rfcs/pull/4), [electron/rfcs#8](https://github.com/electron/rfcs/pull/8)
```

I think the filename also needs to be updated to `0008-preload-realm.md` to match the PR number.

**Review by erickzhao — APPROVED** [review]
RFC LGTM

**Review by jkleinsc — COMMENTED** [review]
RFC LGTM

**Comment by samuelmaddock** [comment]
Regarding ShadowRealm TC39 proposal status, there's been a couple updates this year.
- February: https://github.com/tc39/notes/blob/main/meetings/2024-02/feb-7.md#shadowrealms-update
- June: https://github.com/tc39/notes/blob/main/meetings/2024-06/june-12.md#shadowrealm-update

Before the proposal is able to move forward, more work is needed on [confirming the selection of APIs](https://github.com/tc39/proposal-shadowrealm/issues/393) to be included within the new realm. Also, more test coverage is needed with WTP mentioned.

The good news is that the V8 implementation seems to be stable according to the [February slides](https://docs.google.com/presentation/d/1fd5-VKtl0LxYitLHr_bJ82_xaLs1w5xSmD0dCdh2TvU/edit#slide=id.g266c32110d7_0_0) and [I added a WPT test](https://github.com/web-platform-tests/wpt/commit/302696854a153a3f476c50574aab87685f402f22) as part of this RFC :)

Assuming no major changes regarding Blink's implementation, we may see more built-in APIs exposed within the ShadowRealm which would be beneficial for Preload Realms as well.

**Pull request merged**